pkg/hnsw/hnsw.go

```diff
@@ -148,18 +152,138 @@ func (h *Hnsw) findCloserEntryPoint(q *Point, qFriends *Friends) *Item {
 	return epItem
 }
 
+func (h *Hnsw) selectNeighbors(nearestNeighbors *BaseQueue) (*BaseQueue, error) {
+	maxNearestNeighbors := FromBaseQueue(nearestNeighbors, MaxComparator{})
+
+	for maxNearestNeighbors.Len() > h.M {
+		_, err := maxNearestNeighbors.PopItem()
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return FromBaseQueue(maxNearestNeighbors, MinComparator{}), nil
```

Comment:
Is this necessary? It seems somewhat expensive to reheapify the list every iteration.

Reply:
Yeah, reading through the code I think we can just implicitly pop the closest `h.M` items instead of reheapifying to produce a heap with the correct size. That will save two calls to `FromBaseQueue`